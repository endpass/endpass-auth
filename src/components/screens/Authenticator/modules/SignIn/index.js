export { default } from './SignIn.interface';
