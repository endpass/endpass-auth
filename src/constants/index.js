export const METHODS = {
  SIGN: 'SIGN',
  ACCOUNT: 'ACCOUNT',
  RECOVER: 'RECOVER',
  GET_SETTINGS: 'GET_SETTINGS',

  AUTH: 'AUTH',
  LOGOUT: 'LOGOUT',
  DIALOG_RESIZE: 'DIALOG_RESIZE',
  DIALOG_OPEN: 'DIALOG_OPEN',
  DIALOG_CLOSE: 'DIALOG_CLOSE',
  INITIATE: 'INITIATE',
  READY_STATE_BRIDGE: 'READY_STATE_BRIDGE',
};

export const LS_SETTINGS = 'endpass-connect:settings';

export const IDENTITY_MODE = Object.freeze({
  DEFAULT: 'default',
  CUSTOM: 'custom',
  LOCAL: 'local',
});

export const DIRECTION = Object.freeze({
  AUTH: 'auth',
  CONNECT: 'connect',
});

export const ORIGIN_HOST = (() => {
  const parser = document.createElement('a');
  parser.href = document.referrer;
  return parser.origin;
})();
