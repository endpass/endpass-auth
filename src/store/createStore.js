// @ts-check
import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

/**
 * @param {object?} [options] passed options for store
 * @return {import('vuex').Store<{}>} store Vuex store instance
 */
export default options => {
  return new Vuex.Store({
    strict: !ENV.VUE_APP_IS_PRODUCTION,
    ...options,
  });
}
