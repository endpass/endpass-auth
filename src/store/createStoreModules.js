import GasPriceModule from '@/store/modules/GasPriceModule';
import SharedModule from '@/store/modules/SharedModule';
import ChannelStore from '@/store/modules/ChannelModule';
import BalanceModule from '@/store/modules/BalanceModule';
import AccountsModule from '@/store/modules/AccountsModule';
import AuthModule from '@/store/modules/AuthModule';
import CoreModule from '@/store/modules/CoreModule';
import RequestsModule from '@/store/modules/RequestsModule';
import moduleFactory from '@/store/moduleFactory';

export default store => {
  const createModule = moduleFactory(store);

  const gasPriceStore = createModule(GasPriceModule, 'gasPrice');

  const sharedStore = createModule(SharedModule, 'shared');

  const channelStore = createModule(ChannelStore, 'channel');

  const balanceStore = createModule(BalanceModule, 'balance');

  const accountsStore = createModule(AccountsModule, 'accounts', {
    sharedStore,
    balanceStore,
  });

  const authStore = createModule(AuthModule, 'auth', {
    sharedStore,
  });

  const coreStore = createModule(CoreModule, 'core', {
    authStore,
    sharedStore,
  });

  const requestStore = createModule(RequestsModule, 'requests', {
    accountsStore,
    sharedStore,
  });

  return {
    channelStore,
    gasPriceStore,
    sharedStore,
    accountsStore,
    balanceStore,
    authStore,
    coreStore,
    requestStore,
  };
};
