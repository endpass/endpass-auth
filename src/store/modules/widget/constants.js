export const WIDGET_RESIZE_DURATION = 300;

export default {};
