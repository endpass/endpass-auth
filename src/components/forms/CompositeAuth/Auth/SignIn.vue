<template>
  <form
    data-test="auth-form"
    @submit.prevent="onSignIn"
  >
    <form-item v-if="isServerMode && !isPublic">
      <v-title>
        <span v-html="$t('components.serverModeSelect.identityServer')" />
      </v-title>
      <server-mode-select
        v-model="serverMode"
        @confirm="onSubmit"
      />
      <v-spacer :height="4" />
      <v-divider />
    </form-item>
    <template v-if="isDefaultMode">
      <v-title>
        <span v-html="$t('components.auth.loginToContinue')" />
      </v-title>
      <form-item v-if="error">
        <message
          :error="true"
          data-test="error-message"
        >
          {{ error }}
        </message>
      </form-item>
      <form-item>
        <v-input
          v-model="email"
          v-validate="'required|email'"
          data-vv-as="email"
          data-vv-name="email"
          :error="errors.first('email')"
          name="email"
          type="email"
          :placeholder="$t('components.auth.enterEmail')"
          data-test="email-input"
        />
      </form-item>
      <v-button
        :disabled="!isSubmitEnable"
        :is-loading="isLoading"
        size="big"
        data-test="submit-button-auth"
      >
        {{ $t('global.continue') }}
      </v-button>
      <v-spacer :height="3" />
      <v-divider>{{ $t('components.auth.orSignInWith') }}</v-divider>
      <form-row>
        <form-controls>
          <google-auth-button
            type="button"
            @submit="onSocialSubmit"
            @error="onOauthError"
          />
          <git-auth-button
            type="button"
            @submit="onSocialSubmit"
            @error="onOauthError"
          />
        </form-controls>
      </form-row>
      <v-divider />
      <form-row class="v-text-center v-fw-b">
        {{ $t('components.auth.dontHaveAccount') }}&nbsp;
        <v-link
          href="#"
          data-test="switch-to-sign-up"
          @click.prevent="onSwitch"
        >
          {{ $t('components.auth.signUp') }}
        </v-link>
      </form-row>
    </template>
  </form>
</template>

<script>
import VInput from '@endpass/ui/kit/VInput';
import VButton from '@endpass/ui/kit/VButton';
import VDivider from '@endpass/ui/kit/VDivider';
import VLink from '@endpass/ui/kit/VLink';
import FormItem from '@/components/common/FormItem';
import FormRow from '@/components/common/FormRow';
import VSpacer from '@/components/common/VSpacer';
import ServerModeSelect from '@/components/common/ServerModeSelect';
import GoogleAuthButton from '@/components/common/GoogleAuthButton';
import GitAuthButton from '@/components/common/GitAuthButton';
import Message from '@/components/common/Message';
import { IDENTITY_MODE } from '@/constants';
import formMixin from '@/mixins/form';
import VTitle from '@/components/common/VTitle';
import { coreStore, authStore } from '@/store';
import FormControls from '@/components/common/FormControls';

export default {
  name: 'SignInForm',

  coreStore,
  authStore,

  props: {
    isPublic: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    isLoading: false,
    email: '',
    error: '',
    serverMode: {
      type: IDENTITY_MODE.DEFAULT,
      serverUrl: undefined,
    },
  }),

  computed: {
    isServerMode() {
      return this.$options.coreStore.isServerMode;
    },

    isDefaultMode() {
      return this.serverMode.type === IDENTITY_MODE.DEFAULT;
    },

    isLocalMode() {
      return this.serverMode.type === IDENTITY_MODE.LOCAL;
    },

    isCustomMode() {
      return this.serverMode.type === IDENTITY_MODE.CUSTOM;
    },

    isSubmitEnable() {
      const {
        isDefaultMode,
        isLocalMode,
        isCustomMode,
        isFormValid,
        isLoading,
      } = this;
      const isDefaultValid = isDefaultMode && isFormValid;

      return (isDefaultValid || isCustomMode || isLocalMode) && !isLoading;
    },
  },

  watch: {
    email() {
      this.error = null;
    },
  },

  methods: {
    onSwitch() {
      this.$emit('sign-in');
    },

    onSubmit() {
      if (!this.isSubmitEnable) {
        return;
      }
      this.handleSubmit();
    },

    handleSubmit() {
      const { email, serverMode } = this;

      this.$emit('submit', { email, serverMode });
    },

    async onSignIn() {
      if (!this.isSubmitEnable) {
        return;
      }

      try {
        const { email, serverMode } = this;

        if (serverMode.type !== IDENTITY_MODE.DEFAULT) {
          this.handleSubmit();
          return;
        }

        this.isLoading = true;
        await this.$options.authStore.loadAuthChallenge({ email });

        this.handleSubmit();
      } catch (error) {
        this.error = error;
      } finally {
        this.isLoading = false;
      }
    },

    async onSocialSubmit() {
      this.isLoading = true;
      await this.$options.authStore.waitLogin();
      this.$emit('social');
      this.isLoading = false;
    },

    onOauthError() {
      this.error = this.$i18n.t('components.compositeAuth.authFailed');
    },
  },
  mixins: [formMixin],
  components: {
    FormControls,
    VLink,
    VTitle,
    VButton,
    VInput,
    VSpacer,
    VDivider,
    GoogleAuthButton,
    GitAuthButton,
    Message,
    FormItem,
    FormRow,
    ServerModeSelect,
  },
};
</script>
