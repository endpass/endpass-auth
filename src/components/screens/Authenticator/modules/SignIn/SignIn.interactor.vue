<template>
  <sign-in
    :is-loading="isLoading"
    :error.sync="error"
    :is-server-mode="isServerMode"
    @sign-up="onSignUp"
    @submit="onSignIn"
    @social="onSocial"
  />
</template>

<script>
import { IDENTITY_MODE } from '@/constants';
import { coreStore, authStore } from '@/store';
import SignIn from './SignIn.view';

export default {
  name: 'SignInInteractor',

  coreStore,
  authStore,

  data: () => ({
    isLoading: false,
    error: '',
  }),

  computed: {
    isServerMode() {
      return this.$options.coreStore.isServerMode;
    },
  },

  methods: {
    onSignUp() {
      this.$emit('sign-up');
    },

    async onSignIn({ email, serverMode }) {
      if (this.isLoading) return;

      try {
        this.isLoading = true;
        this.error = '';

        let isPasswordExist = false;

        if (serverMode.type === IDENTITY_MODE.DEFAULT) {
          await this.$options.authStore.loadAuthChallenge({ email });

          isPasswordExist = await this.$options.authStore.checkRegularPassword(
            email,
          );
        }

        this.$emit('sign-in', { email, serverMode, isPasswordExist });
      } catch (error) {
        this.error = this.$i18n.t('components.compositeAuth.authFailed');
      } finally {
        this.isLoading = false;
      }
    },

    async onSocial() {
      this.isLoading = true;

      await this.$options.authStore.waitLogin();
      this.$emit('social');

      this.isLoading = false;
    },
  },

  components: {
    SignIn,
  },
};
</script>
