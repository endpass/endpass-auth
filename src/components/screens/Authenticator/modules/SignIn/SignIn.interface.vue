<template>
  <sign-in
    v-bind="$attrs"
    :is-public="isPublic"
    @sign-in="onSignIn"
    @sign-up="onSignUp"
    @social="onSocial"
  />
</template>

<script>
import SignIn from './SignIn.interactor';

export default {
  name: 'SignInInterface',

  props: {
    isPublic: {
      type: Boolean,
      default: false,
    },
  },

  methods: {
    onSignUp() {
      this.$emit('switch', { to: 'sign-up' });
    },

    onSignIn({ email, serverMode, isPasswordExist, isSignUp = false }) {
      this.$emit('complete', { email, serverMode, isPasswordExist, isSignUp });
    },

    onSocial() {
      this.$emit('social');
    },
  },

  components: {
    SignIn,
  },
};
</script>
