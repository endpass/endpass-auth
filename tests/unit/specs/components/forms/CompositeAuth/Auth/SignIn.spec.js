import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import VeeValidate from 'vee-validate';
import SignIn from '@/components/forms/CompositeAuth/Auth/SignIn';
import { IDENTITY_MODE } from '@/constants';
import setupI18n from '@/locales/i18nSetup';
import createStore from '@/store/createStore';
import createStoreModules from '@/store/createStoreModules';
import identityService from '@/service/identity';

describe('SignIn', () => {
  let wrapper;
  let coreStore;

  const localVue = createLocalVue();
  localVue.use(VeeValidate);
  const i18n = setupI18n(localVue);

  const createWrapper = (options = {}) => {
    const store = createStore();
    const { coreStore: coreStoreModule } = createStoreModules(store);

    coreStore = coreStoreModule;

    return shallowMount(SignIn, {
      localVue,
      coreStore,
      propsData: {
        isPublic: false,
      },
      provide: {
        theme: 'default',
      },
      i18n,
      sync: false,
      ...options,
    });
  };

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = createWrapper();
  });

  describe('render', () => {
    it('should correctly render SignIn component', () => {
      expect(wrapper.name()).toBe('SignInForm');
      expect(wrapper.find('[data-test=email-input]').exists()).toBe(true);
      expect(wrapper.html()).toMatchSnapshot();
    });

    it.skip('should render error', () => {
      // TODO: add check for show error for identity request in other tests
      wrapper = createWrapper({
        propsData: {
          message: 'foo',
          error: 'bar',
        },
      });

      expect(wrapper.findAll('[data-test=error-message]').exists()).toBe(true);
    });

    it.skip('should change submit button text if loading and make it disabled', () => {
      wrapper = createWrapper();
      wrapper.setData({
        isLoading: true,
      });

      const submitButton = wrapper.find('[data-test=submit-button-auth]');

      expect(wrapper.html()).toMatchSnapshot();
      expect(submitButton.text()).toBe('Loading...');
      expect(submitButton.attributes().disabled).toBe('true');
    });

    describe('server mode', () => {
      it('should not render server mode component by default', () => {
        expect(wrapper.find('server-mode-select-stub').exists()).toBe(false);
      });

      it('should render server mode component', async () => {
        expect.assertions(2);

        expect(wrapper.find('server-mode-select-stub').exists()).toBe(false);

        coreStore.isServerMode = true;

        await wrapper.vm.$nextTick();

        expect(wrapper.find('server-mode-select-stub').exists()).toBe(true);
      });

      it.skip('should not call auth if mode is not default', async () => {
        expect.assertions(2);

        wrapper.find('form-stub').vm.$emit('submit', {
          email: 'email',
          serverMode: { type: IDENTITY_MODE.CUSTOM },
        });

        await global.flushPromises();

        expect(identityService.auth).not.toBeCalled();

        expect(wrapper.emitted().authorize[0]).toEqual([
          { serverMode: { type: 'custom' } },
        ]);
      });
    });
  });

  describe('behavior', () => {
    const email = 'foo+@bar.baz';
    const defaultServerMode = {
      type: IDENTITY_MODE.DEFAULT,
      serverUrl: undefined,
    };
    const defaultEmitParams = { email, serverMode: defaultServerMode };

    describe('form', () => {
      beforeEach(() => {
        wrapper = mount(SignIn, {
          localVue,
          provide: {
            theme: 'default',
          },
          i18n,
          sync: false,
        });
      });

      describe('email', () => {
        describe('empty', () => {
          it('should not allow submit form', () => {
            wrapper.find('form').trigger('submit');

            expect(wrapper.emitted().submit).toBe(undefined);
          });
        });

        describe('invalid', () => {
          it('should disable submit button', () => {
            ['foo@bar', 'foo foo@bar.com', 'foo@bar.com foo'].forEach(
              badEmail => {
                wrapper.setData({
                  email: badEmail,
                });

                expect(
                  wrapper.find('[data-test=submit-button-auth]').attributes()
                    .disabled,
                ).toBe('disabled');
              },
            );
          });

          it('should not allow submit form', async () => {
            expect.assertions(1);

            wrapper.setData({
              email: 'foo@bar',
            });
            wrapper.find('form').trigger('submit');

            await global.flushPromises();

            expect(wrapper.emitted().submit).toBe(undefined);
          });
        });

        describe('valid', () => {
          beforeEach(() => {
            wrapper.setData({
              email,
            });
          });

          it('should enable submit button', async () => {
            expect.assertions(1);

            await global.flushPromises();

            expect(
              wrapper.find('[data-test=submit-button-auth]').attributes()
                .disabled,
            ).not.toBe('disabled');
          });

          it.skip('should allow submit', () => {
            wrapper.find('form').trigger('submit');

            expect(wrapper.emitted().submit).toEqual([[defaultEmitParams]]);
          });
        });
      });
    });

    describe('server mode', () => {
      beforeEach(() => {
        coreStore.isServerMode = true;
      });

      describe('emit', () => {
        it('should not render email form when custom event', async () => {
          expect.assertions(1);

          const customServerMode = {
            type: IDENTITY_MODE.CUSTOM,
            serverUrl: 'https://site.com',
          };

          wrapper
            .find('server-mode-select-stub')
            .vm.$emit('input', customServerMode);

          await wrapper.vm.$nextTick();

          expect(wrapper.find('[data-test=email-input]').exists()).toBe(false);
        });

        it('should not render email form when local event', async () => {
          expect.assertions(1);

          const localServerMode = {
            type: IDENTITY_MODE.LOCAL,
            serverUrl: undefined,
          };

          wrapper
            .find('server-mode-select-stub')
            .vm.$emit('input', localServerMode);

          await wrapper.vm.$nextTick();

          expect(wrapper.find('[data-test=email-input]').exists()).toBe(false);
        });

        it('should emit submit event when confirm', async () => {
          expect.assertions(1);

          wrapper.setData({
            email,
          });
          await global.flushPromises();

          wrapper.find('server-mode-select-stub').vm.$emit('confirm');
          await wrapper.vm.$nextTick();

          expect(wrapper.emitted().submit).toEqual([[defaultEmitParams]]);
        });
      });
    });
  });
});
