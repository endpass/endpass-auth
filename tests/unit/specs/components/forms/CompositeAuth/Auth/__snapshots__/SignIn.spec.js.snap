// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn render should change submit button text if loading and make it disabled 1`] = `
<form data-test="auth-form">
  <!---->
  <v-title-stub><span>Log in&nbsp;to&nbsp;your Endpass account to&nbsp;access site actions</span></v-title-stub>
  <!---->
  <form-item-stub>
    <v-input-stub value="" data-vv-as="email" data-vv-name="email" name="email" type="email" placeholder="Enter your email..." data-test="email-input"></v-input-stub>
  </form-item-stub>
  <v-button-stub skin="primary" size="big" data-test="submit-button-auth">
    Continue
  </v-button-stub>
  <v-spacer-stub height="3" width="0"></v-spacer-stub>
  <v-divider-stub>or sign in with</v-divider-stub>
  <form-row-stub>
    <form-controls-stub>
      <google-auth-button-stub type="button"></google-auth-button-stub>
      <v-spacer-stub height="0" width="16"></v-spacer-stub>
      <git-auth-button-stub type="button"></git-auth-button-stub>
    </form-controls-stub>
  </form-row-stub>
  <v-divider-stub></v-divider-stub>
  <form-row-stub centered="" bold="">
    Don't have an account?&nbsp;
    <v-link-stub href="#" data-test="switch-to-sign-up">
      Sign up
    </v-link-stub>
  </form-row-stub>
</form>
`;

exports[`SignIn render should correctly render SignIn component 1`] = `
<form data-test="auth-form">
  <!---->
  <v-title-stub><span>Log in&nbsp;to&nbsp;your Endpass account to&nbsp;access site actions</span></v-title-stub>
  <!---->
  <form-item-stub>
    <v-input-stub value="" data-vv-as="email" data-vv-name="email" name="email" type="email" placeholder="Enter your email..." data-test="email-input"></v-input-stub>
  </form-item-stub>
  <v-button-stub skin="primary" size="big" data-test="submit-button-auth" disabled="true">
    Continue
  </v-button-stub>
  <v-spacer-stub height="3" width="0"></v-spacer-stub>
  <v-divider-stub>or sign in with</v-divider-stub>
  <form-row-stub>
    <form-controls-stub>
      <google-auth-button-stub type="button"></google-auth-button-stub>
      <git-auth-button-stub type="button"></git-auth-button-stub>
    </form-controls-stub>
  </form-row-stub>
  <v-divider-stub></v-divider-stub>
  <form-row-stub class="v-text-center v-fw-b">
    Don't have an account?&nbsp;
    <v-link-stub href="#" data-test="switch-to-sign-up">
      Sign up
    </v-link-stub>
  </form-row-stub>
</form>
`;
